File: prioritarr/clients/tautulli.py
from __future__ import annotations

import httpx


class TautulliClient:
    """Thin synchronous wrapper around the Tautulli v2 REST API."""

    def __init__(self, base_url: str, api_key: str, timeout: float = 120.0) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._client = httpx.Client(timeout=timeout)

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _call(self, cmd: str, **params: object) -> object:
        query = {"apikey": self._api_key, "cmd": cmd, **params}
        response = self._client.get(f"{self._base_url}/api/v2", params=query)
        response.raise_for_status()
        return response.json()["response"]["data"]

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def get_show_libraries(self) -> list:
        data = self._call("get_libraries")
        return [lib for lib in data if lib.get("section_type") == "show"]

    def get_history(
        self,
        grandparent_rating_key: str | None = None,
        media_type: str | None = None,
        length: int = 1000,
    ) -> list:
        kwargs: dict[str, object] = {"length": length}
        if grandparent_rating_key is not None:
            kwargs["grandparent_rating_key"] = grandparent_rating_key
        if media_type is not None:
            kwargs["media_type"] = media_type
        data = self._call("get_history", **kwargs)
        return data["data"]

    def get_metadata(self, rating_key: str) -> object:
        return self._call("get_metadata", rating_key=rating_key)

    def get_library_media_info(self, section_id: int, length: int = 5000) -> list:
        data = self._call("get_library_media_info", section_id=section_id, length=length)
        return data["data"]

File: prioritarr/backend/src/main/kotlin/org/yoshiz/app/prioritarr/backend/clients/Tautulli.kt
package org.yoshiz.app.prioritarr.backend.clients

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.get
import io.ktor.client.request.parameter
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject

/** Thin wrapper around the Tautulli v2 API. Mirrors prioritarr/clients/tautulli.py. */
class TautulliClient(
    private val baseUrl: String,
    private val apiKey: String,
    private val http: HttpClient,
) {
    private val root: String = baseUrl.trimEnd('/')

    suspend fun getShowLibraries(): JsonArray {
        val data = call("get_libraries") as JsonArray
        return JsonArray(data.filter {
            (it.jsonObject["section_type"]?.jsonPrimitiveOrNull?.contentOrNull == "show")
        })
    }

    suspend fun getHistory(
        grandparentRatingKey: String? = null,
        mediaType: String? = null,
        length: Int = 1000,
    ): JsonArray {
        val params = mutableMapOf<String, String>("length" to length.toString())
        grandparentRatingKey?.let { params["grandparent_rating_key"] = it }
        mediaType?.let { params["media_type"] = it }
        val data = call("get_history", params) as JsonObject
        return data["data"] as JsonArray
    }

    suspend fun getMetadata(ratingKey: String): JsonElement =
        call("get_metadata", mapOf("rating_key" to ratingKey))

    suspend fun getLibraryMediaInfo(sectionId: Int, length: Int = 5000): JsonArray {
        val data = call(
            "get_library_media_info",
            mapOf("section_id" to sectionId.toString(), "length" to length.toString()),
        ) as JsonObject
        return data["data"] as JsonArray
    }

    private suspend fun call(cmd: String, params: Map<String, String> = emptyMap()): JsonElement {
        val response: JsonObject = http.get("$root/api/v2") {
            parameter("apikey", apiKey)
            parameter("cmd", cmd)
            for ((k, v) in params) parameter(k, v)
        }.body()
        // Tautulli wraps everything in {"response":{"result":"success","data": ...}}
        return (response["response"] as JsonObject)["data"]
            ?: error("Tautulli cmd=$cmd returned no data field")
    }
}

private val JsonElement.jsonPrimitiveOrNull
    get() = this as? kotlinx.serialization.json.JsonPrimitive
private val kotlinx.serialization.json.JsonPrimitive.contentOrNull: String?
    get() = if (isString) content else content
